### ISSUE TITLE
dap_bind(): Unable to bind to server: Can't contact LDAP server with Active Directory Enterprise CA root certificate

### ISSUE THREAD

**Issue by stschatzl** [ltb-project/self-service-password#596]
Hello,

I cannot get the password change working with an Active Directory domain which has LDAPS (with a standalone Enterprise CA) running.

On submitting the change, it returns that there is no contact possible, if LDAPS (SSL) is activated, although I injected the exported Windows Active Directory Enterprise CA root certificate and ran update-ca-certificates on docker build.

If I contact the SAME LDAP server ( = domain controller) WITHOUT SSL, it can bind to it and submits the password change which is blocked by the Active Directory, as LDAPS is enforced for password changes. I can also verify that this is not a network problem or something, as I can do on the same (Linux) machine with Apache Directory Studio anything with LDAPS on the same LDAPS server without a hitch after accepting the Enterprise CA ssl root certificate.

On windows exported ca certificate with:
```
certutil -ca.cert contoso-ca-certroot.crt
CA cert[0]: 3 -- Valid
CA cert[0]:
-----BEGIN CERTIFICATE-----
MIIFnTCCA4WgAwIBAgIQNuLXKHPBMLZDV2J+DKxU6DANBgkqhkiG9w0BAQsFADBU
...
Rw==
-----END CERTIFICATE-----

CertUtil: -ca.cert command completed successfully.
```

config.inc.php:
```
# LDAP
#$ldap_url = "ldap://server.contoso.com";
$ldap_url = "ldaps://server.contoso.com";
$ldap_starttls = false;
$ldap_binddn = "cn=ldapuser, cn=Users,DC=contoso, DC=com";
$ldap_bindpw = 'the-secret-password';
// for GSSAPI authentication, comment out ldap_bind* and uncomment ldap_krb5ccname lines
//$ldap_krb5ccname = "/path/to/krb5cc";
$ldap_base = "cn=Users,DC=contoso, DC=com";
$ldap_login_attribute = "uid";
$ldap_fullname_attribute = "cn";
// openldap: $ldap_filter = "(&(objectClass=person)($ldap_login_attribute={login}))";
// AD
$ldap_filter = "(&(objectClass=user)(sAMAccountName={login})(!(userAccountControl:1.2.840.113556.1.4.803:=2)))";
$ldap_use_exop_passwd = false;
$ldap_use_ppolicy_control = false;

# Active Directory mode
# true: use unicodePwd as password field
# false: LDAPv3 standard behavior
$ad_mode = true;
# Force account unlock when password is changed
$ad_options['force_unlock'] = true;
# Force user change password at next login
$ad_options['force_pwd_change'] = false;
# Allow user with expired password to change password
$ad_options['change_expired_password'] = true;
```

I added the following to the Dockerfile for building:
```
# allow more memory
RUN sed -i 's/memory_limit\ \=\ 128M/memory_limit\ \=\ 512M/g' "$PHP_INI_DIR/php.ini"
# update SSL certificate
COPY conf/contoso-ca-certroot.crt /usr/local/share/ca-certificates/
RUN /usr/sbin/update-ca-certificates
```

```
docker build -t docker-self-service-password .
....
---> 1a2a0a6adaee
Step 7/11 : RUN sed -i 's/memory_limit\ \=\ 128M/memory_limit\ \=\ 512M/g' "$PHP_INI_DIR/php.ini"
 ---> Running in 20a36263a2bf
Removing intermediate container 20a36263a2bf
 ---> faec942016e1
Step 8/11 : COPY conf/connex-ca-certroot.crt /usr/local/share/ca-certificates/
 ---> 88ebb770d619
Step 9/11 : RUN /usr/sbin/update-ca-certificates
 ---> Running in 4e193320bce7
Updating certificates in /etc/ssl/certs...
1 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Removing intermediate container 4e193320bce7
 ---> f009ccebdf81
Step 10/11 : ENV LC_CTYPE=en_US.UTF-8
 ---> Running in 25c67d1082ce
Removing intermediate container 25c67d1082ce
 ---> 4a943e0aa033
Step 11/11 : EXPOSE 80
 ---> Running in 93fa580b5932
Removing intermediate container 93fa580b5932
 ---> 350e779f436f
Successfully built 350e779f436f
```

Docker was run with these parameters:
```
docker run -p 7777:80 -v /home/dafa/Documents/projects/self-service-pwd-reset/conf/config-contoso.inc.php:/var/www/conf/config.inc.php -it docker-self-service-password-cx:latest
```

After submitting the password change, the docker log gives this error message:
```
[Fri Oct 15 09:18:21.923063 2021] [php7:warn] [pid 18] [client 192.168.2.130:59890] PHP Warning:  ldap_bind(): Unable to bind to server: Can't contact LDAP server in /var/www/html/change.php on line 91, referer: http://192.168.2.130:7777/
[Fri Oct 15 09:18:21.923105 2021] [php7:notice] [pid 18] [client 192.168.2.30:59890] LDAP - Bind error -1  (Can't contact LDAP server), referer: http://192.168.2.130:7777/
[Fri Oct 15 09:18:21.925891 2021] [php7:notice] [pid 18] [client 192.168.2.130:59890] PHP Notice:  Undefined index: msg_changehelpextramessage in /var/www/templates_c/3ba7131c51edfe0cef48341f1409d3c1a60d3c16_2.file.change.tpl.php on line 43, referer: http://192.168.2.130:7777/
```

I tried to adding according to https://stackoverflow.com/questions/3866406/need-help-ignoring-server-certificate-while-binding-to-ldap-server-using-php the line 
```
putenv('LDAPTLS_REQCERT=never');
```
but that did not work.

Any help and suggestions?

**Comment by faust64** [comment]
Are we sure this is related to TLS / have you been working with other configurations before?

From your container:

- can you establish a connection to your LDAP (telnet may be missing, a curl http://ldapaddr:ldapport should at least confirm there's some protocol error, rather than a timeout)
- can you confirm tls verification works (openssl s_client -connect ldapaddr:ldapport)
- have you tried to edit your /etc/ldap/ldap.conf, the TLS_CACERT especially (?)

There's a lot of issues already regarding TLS and active directories, have you searched: https://github.com/ltb-project/self-service-password/issues?q=is%3Aissue+tls

**Comment by shakeyourbunny** [comment]
Concerning the "regular" LDAP connection: unencrypted LDAP connections work fine, also requests to change the password (this is the point where my Active Directory won't process it due the lack of encryption).
```
[Fri Oct 15 09:00:29.082085 2021] [php7:warn] [pid 20] [client 192.168.2.130:59874] PHP Warning:  ldap_modify_batch(): Batch Modify: Server is unwilling to perform in /var/www/lib/functions.inc.php on line 454, referer: http://192.168.2.130:7777/
[Fri Oct 15 09:00:29.082190 2021] [php7:notice] [pid 20] [client 192.168.2.130:59874] LDAP - Modify password error 53 (Server is unwilling to perform), referer: http://192.168.2.130:7777/
```

LDAP SSL connection works within the container with the added certificates.
```
root@cd786a10e99f:/var/www/html# openssl s_client -verify_return_error -connect server.contoso.com:636
CONNECTED(00000003)
depth=1 DC = com, DC = contoso, CN = ContosoEnterpriseCA
verify return:1
depth=0 CN = server.contoso.com
verify return:1
---
Certificate chain
 0 s:CN = server.contoso.com
   i:DC = com, DC = contoso, CN = ContosoEnterpriseCA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIHGjCCBQKgAwIBAgITHQAAAAORPJX91ldD8wAAAAAAAzANBgkqhkiG9w0BAQsF
ADBUMRMwEQYKCZImiZPyLGQBGRYDbmV0MRswGQYKCZImiZPyLGQBGRYLY29ubmV4
....
9VjvluFFMtUht5Homzi7PxSczF796LmekbGkfmvXZZIM2X4RDXLF/A1O+j7Uax5M
4xHLtpxWC9p4zeRBHHva8oTF8l8fiHGztKhnsCL2FgoBZryaj86UQt/efqkrpWck
DMQYUGJKqwqGCj1APHdCaVt9xwqIPx+AUF4KGR/eLALEOaCCy0aPGMYJRY4etA==
-----END CERTIFICATE-----

subject=CN = server.contoso.com

issuer=DC = com, DC = contoso, CN = ContosoEnterpriseCA

---
No client certificate CA names sent
Client Certificate Types: RSA sign, DSA sign, ECDSA sign
Requested Signature Algorithms: RSA+SHA256:RSA+SHA384:RSA+SHA1:ECDSA+SHA256:ECDSA+SHA384:ECDSA+SHA1:DSA+SHA1:RSA+SHA512:ECDSA+SHA512
Shared Requested Signature Algorithms: RSA+SHA256:RSA+SHA384:ECDSA+SHA256:ECDSA+SHA384:RSA+SHA512:ECDSA+SHA512
Peer signing digest: SHA256
Peer signature type: RSA
Server Temp Key: ECDH, P-384, 384 bits
---
SSL handshake has read 2372 bytes and written 487 bytes
Verification: OK
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 73100000EFD96B133C26BCB623D1BB53864213B9D77212246D50182398EFDEB2
    Session-ID-ctx: 
    Master-Key: 4996AC075BE335EE16EBA1072B772715DE76C1FA82B7765D3CE96E49F9B01C50A44607B0753E0D83A38EB16628ED843B
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1634299780
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: yes
---
```

Regarding the container (built from the Dockerfile in the github repo) there is no /etc/ldap/openldap.conf present?

I already skimmed through the problems on https://github.com/ltb-project/self-service-password/issues?q=is%3Aissue+tls, but there was not really something what helped.

Adding $debug = true will report, when I'm back at work.

In General (and out of scope of this project) I also verified that this is not a connection problem or a firewall sitting between which is blocking, because I have a LDAP editor called "Apache Directory Studio" installed which can connect with LDAP and LDAP to the same server fine after accepting the certificate ( -> https://directory.apache.org/studio/)

**Comment by shakeyourbunny** [comment]
Addendum: it seems that the Debian 11 container (Dockerfile from repo) can do LDAPS connections, but not the running Apache / PHP script. It is really strange.

On the Active Directory side, the Active Directory server is a fully patched (as of October 2021) Windows Server 2019.

**Comment by faust64** [comment]
I'm pretty sure I've seen another case with similar symptoms: ldapsearch works, but php doesn't, ... I remember we tried adding a `.ldaprc` file into the webserver root. I can't recall how the reporting user fixed it. Though it was most likely in this repository, or the https://github.com/ltb-project/white-pages/issues one ...
